### DISCUSSION TITLE
How to customize indent size?

### DISCUSSION THREAD

**Discussion by markoa** [zed-industries/zed#6586] [Feature Requests]
Self explanatory...

I'm used to Vim where I have my global setting (2 ftw) which specific language packs can override.

**Comment by JosephTLyons** [comment]
Hey @markoa, you can specify the indent setting used by all languages by placing this at the root of your `settings.json` file:

```json
{
  "tab_size": 4,
}
```

If you want to override this setting on a per-language basis, you can use something like this:

```json
{
  "tab_size": 4,
  "language_overrides": {
    "Python": {
      "tab_size": 2
    }
  }
}
```

Note that these are still considered gobal settings and you can override these settings on a per-project basis as well, by adding a `.zed/settings.json` to your project.  There is a command palette action that can create this for you as well: `zed: open local settings`.

**Reply by markoa** [reply]
Great, thank you @JosephTLyons!

**Reply by sshine** [reply]
I don't know if this changed over time, but [the current docs](https://zed.dev/docs/configuring-zed) (2024-07-23) suggest:

```
{
  "tab_size": 4,
  "languages": {
    "Python": {
      "tab_size": 2
    }
  }
}
```

I.e. still [`tab_size`](https://zed.dev/docs/configuring-zed#tab-size), but [`languages`](https://zed.dev/docs/configuring-zed#languages) instead of `language_overrides`.

**Reply by the2nake** [reply]
> If you want to override this setting on a per-language basis, you can use something like this:
> 
> ```json
> {
>   "tab_size": 4,
>   "language_overrides": {
>     "Python": {
>       "tab_size": 2
>     }
>   }
> }
> ```

Could you update this answer to reflect that "language_overrides" has changed to "languages"?

**Comment by SuperKenVery** [comment]
Is it possible to make zed auto detect the current file's indent?

**Reply by weebao** [reply]
Need this actually

**Comment by dushmanta05** [comment]
It was helpful. Thank you.

**Comment by swznd** [comment]
how to change indent size individual file ?

**Reply by SuperKenVery** [reply]
I believe zed has "project settings" which you could open in the command pallette. That should help.

**Reply by swznd** [reply]
I mean for individual files, not entire projects or languages. some legacy code have different indent sizes for different modules

**Comment by vixalien** [comment]
It seems when you format a file, it doesn't format the indent_size you configured

**Reply by d3vma** [reply]
yup! having `tab_size` in both `settings.json` and project specific settings ain't reflect nothing!

**Reply by Coenicorn** [reply]
Has this been fixed, or at least addressed in another setting? I can't seem to stop zed from implying the tab size from the current document instead of explicitly using the space specified in `settings.json`

**Comment by dynamsoft-h** [comment]
How do I make markdown use even-numbered indent spaces for nested lists? It currently uses 3, 5, 7, etc. at each level, but I want it to use 2, 4, 6, etc. I have the settings ` "tab_size": 2` and `"hard_tabs": false`.

**Comment by Kiryonn** [comment]
I tried to use global settings, project settings and even .editorconfig, nothing let me use tabs as indents or change the indent size.

**Reply by Kiryonn** [reply]
I'v found the solution if you use ruff for python.
You have to create `pyproject.toml` at the root of your project and write
```py
[tool.ruff]
line-length = 108

[tool.ruff.format]
indent-style = "tab"
quote-style = "double"
```

**Comment by iwinux** [comment]
Wow it's 2026 and Zed can't even adjust indent size for individual files.

**Reply by weebao** [reply]
Wonder if there's an issue for this

**Reply by chulkilee** [reply]
Per-file tab setting -> #4867

**Comment by jb-alvarado** [comment]
If the formatter is prettier, then it can be configured with a **.prettierrc** file, like this:

```JSON
{
    "tabWidth": 4,
    "overrides": [
        {
            "files": "package*.json",
            "options": {
                "tabWidth": 2
            }
        }
    ]
}
```

Check https://prettier.io/docs/configuration for more option.
